lib/teleterm/services/connectmycomputer/connectmycomputer.go

```diff
@@ -0,0 +1,255 @@
+	err := cfg.CheckAndSetDefaults()
+	if err != nil {
+		return nil, err
+	}
+
+	return &RoleSetup{cfg: cfg}, nil
+}
+
+type RoleSetupResult struct {
+	CertsReloaded bool
+}
+
+// Run ensures that at the end of its execution the user has their own individual Connect My
+// Computer role and that the role includes the current system username in allowed logins.
+//
+// If the role list of the user got updated, the return value has CertsReloaded set to true.
+func (s *RoleSetup) Run(ctx context.Context, accessAndIdentity AccessAndIdentity, certManager CertManager, cluster *clusters.Cluster) (RoleSetupResult, error) {
+	noCertsReloaded := RoleSetupResult{}
+	if !cluster.URI.IsRoot() {
+		return noCertsReloaded, trace.BadParameter("Connect My Computer works only with root clusters")
+	}
+
+	// Do not use GetCurrentUser – it returns the current view of the user given the certs, not merely
+	// the resource from the backend. This means that GetCurrentUser includes any roles granted
+	// through access requests. We don't want that since we're later going to update the user.
+	clusterUser, err := accessAndIdentity.GetUser(cluster.GetLoggedInUser().Name, false /* withSecrets */)
+	if err != nil {
+		return noCertsReloaded, trace.Wrap(err)
+	}
+
+	isLocalUser := clusterUser.GetCreatedBy().Connector == nil
+	if !isLocalUser {
+		return noCertsReloaded,
+			trace.BadParameter("Connect My Computer works only with local users, the user %v was created by %v connector",
+				clusterUser.GetName(), clusterUser.GetCreatedBy().Connector.Type)
+	}
+
+	roleName := fmt.Sprintf("%v%v", teleport.ConnectMyComputerRoleNamePrefix, clusterUser.GetName())
+
+	doesRoleExist := true
+	existingRole, err := accessAndIdentity.GetRole(ctx, roleName)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			doesRoleExist = false
+		} else {
+			return noCertsReloaded, trace.Wrap(err)
+		}
+	}
+
+	systemUser, err := user.Current()
+	if err != nil {
+		return noCertsReloaded, trace.Wrap(err)
+	}
+
+	if !doesRoleExist {
+		s.cfg.Log.Infof("Creating the role %v.", roleName)
+
+		role, err := types.NewRole(roleName, types.RoleSpecV6{
+			Allow: types.RoleConditions{
+				NodeLabels: types.Labels{
+					types.ConnectMyComputerNodeOwnerLabel: []string{clusterUser.GetName()},
+				},
+				Logins: []string{systemUser.Username},
+			},
+		})
+		if err != nil {
+			return noCertsReloaded, trace.Wrap(err)
+		}
+		if err = accessAndIdentity.UpsertRole(ctx, role); err != nil {
+			return noCertsReloaded, trace.Wrap(err, "creating role %v", role.GetName())
+		}
+	} else {
+		if slices.Contains(clusterUser.GetLogins(), systemUser.Username) {
+			s.cfg.Log.Infof("The role %v already exists and includes current system username.", roleName)
+		} else {
+			s.cfg.Log.Infof("Adding %v to the logins of the role %v.", systemUser.Username, roleName)
+
+			allowedLogins := existingRole.GetLogins(types.Allow)
+			existingRole.SetLogins(types.Allow, append(allowedLogins, systemUser.Username))
+
```

Comment:
I forgot to explain why we make sure that the current system username is in allowed logins. I'll add a comment in code about it.

This is to cover a use case where you use Connect My Computer as a single cluster user but from two separate devices. Those devices might use a different system username. Since the role is scoped per cluster user, we need to make sure that the role includes both system usernames in allowed logins.

Now, cluster users are not really supposed to modify this role by themselves. If it was possible to somehow make it readonly for anyone but Connect then I'd do so. And it's a weird one because we can both detect what is wrong with the role _and_ fix it for the user at the same time.

One one hand I hate when a program tells me to do something that it can do by itself. On the other hand, how far do we go with handling possible edge cases?

In this case, I feel like performing the check for the `NodeOwnerLabel` has low enough cost that I think I can add that. Also, we might want to enforce that `connect-my-computer-alice@example.com` gives access only to nodes with the label `owner: alice`, so that's another argument for checking the label here.